samza-core/src/main/java/org/apache/samza/application/ApplicationDescriptorImpl.java

```diff
@@ -176,4 +179,53 @@ public ProcessorLifecycleListenerFactory getProcessorLifecycleListenerFactory()
    */
   public abstract Set<SystemDescriptor> getSystemDescriptors();
 
+  /**
+   * Get all the unique input streamIds in this application
+   *
+   * @return an immutable set of input streamIds
+   */
+  public abstract Set<String> getInputStreamIds();
```

Comment:
This is not necessary if we unify the input streams set between high and low-level applications. Currently, the full set of input stream IDs are stored differently in high and low- level APIs: a) in TaskApplicationDescriptorImpl, getInputDescriptor() will give you the full set of input streams, including the intermediate streams; b) in StreamApplicationDescriptorImpl, getInputOperators() give you the full set of input streams, including the intermediate streams. If we unify the representation of intermediate streams in TaskApplicationDescriptorImpl and StreamApplicationDescriptorImpl, we can remove this method. Before we do that, we need two implementations of the same method in the base class to get the consistent set of input streamIds. I would prefer to open another ticket for that, since this is also only in the Impl classes, not the public API interface class.

Reply:
Minor: In the mean time, it might be helpful to change the internal API such that instead of getting the entire Map of input descriptors and input operators, we only expose a getInputDescriptor(String streamId) and getInputOperator(String streamId). This way it's clearer that the getInputStreamIds is the method that should be used for iteration.